Frameworks/Foundation/NSOperation.mm

```diff
@@ -131,139 +128,109 @@ - (void)setThreadPriority:(double)priority {
-        if (!priv->cancelled) {
-            [self willChangeValueForKey:@"isExecuting"];
-            priv->executing = 1;
-            [self didChangeValueForKey:@"isExecuting"];
-            execute = true;
-        }
-        pthread_mutex_unlock(&priv->finishLock);
-        if (execute) {
-            [self main];
-        }
-
-        if (execute) {
-            [self willChangeValueForKey:@"isExecuting"];
-        }
+    if (_finished) {
+        return;
+    }
 
-        [self _setFinished:YES
-            andPerformUnderLock:^{
-                if (execute) {
-                    priv->executing = 0;
-                    [self didChangeValueForKey:@"isExecuting"];
-                }
-            }];
+    THROW_NS_IF(E_INVALIDARG, (_executing || ![self isReady]));
+    
+    pthread_mutex_lock(&_finishLock);
+    BOOL shouldExecute = !_cancelled; // Note: in the cancelled case, [self main] is not called but isFinished will still be observable
+    if (shouldExecute) {
+        [self willChangeValueForKey:@"isExecuting"];
+        _executing = YES;
+        [self didChangeValueForKey:@"isExecuting"];
+    }
+    pthread_mutex_unlock(&_finishLock);
+    
+    if (shouldExecute) {
+        NSAutoreleasePool* pool = [NSAutoreleasePool new];
+        [self main];
+        [pool release];
     }
-}
 
-/**
- @Status Interoperable
-*/
-- (void)setFinished:(BOOL)finished {
-    [self _setFinished:finished andPerformUnderLock:nil];
+    [self _finish:shouldExecute];
 }
 
-- (void)_setFinished:(BOOL)finished andPerformUnderLock:(void (^)())block {
-    // yes, this is ugly: priv->finished is an int though.
-    int newValue = finished ? 1 : 0;
-    pthread_mutex_lock(&priv->finishLock);
-
-    [self willChangeValueForKey:@"isFinished"];
-
-    if (priv->finished != newValue) {
-        priv->finished = newValue;
+- (void)_finish:(BOOL)didExecute {
+    pthread_mutex_lock(&_finishLock);
+    if (didExecute) {
+        [self willChangeValueForKey:@"isExecuting"];
     }
 
+    [self willChangeValueForKey:@"isFinished"];
+    _finished = YES;
     [self didChangeValueForKey:@"isFinished"];
 
-    if (block) {
-        block();
+    if (didExecute) {
+        _executing = NO;
+        [self didChangeValueForKey:@"isExecuting"];
     }
 
-    if (newValue == 1) {
-        pthread_cond_broadcast(&priv->finishCondition);
-    }
+    pthread_mutex_lock(&_completionBlockLock);
+    void (^completion)(void) = [_completionBlock retain];
+    pthread_mutex_unlock(&_completionBlockLock);
 
-    pthread_mutex_unlock(&priv->finishLock);
-
-    // This may seem unintuitive, but the completion
-    // block is intended to be called even if the operation
-    // is cancelled.
-    if (newValue == 1 && priv->completionBlock) {
-        priv->completionBlock();
-        [priv->completionBlock release];
-        priv->completionBlock = nil;
+    pthread_mutex_unlock(&_finishLock);
+    
+    if (completion) {
+        completion();
     }
+
+    [completion release];
```

Comment:
Sorry... missed the fact this is a local block. Thanks for explaining John

---
In reply to: [69352631](https://github.com/Microsoft/WinObjC/pull/602#discussion_r69352631) [](ancestors = 69352631)